### PR TITLE
Adding Deployment Details Page

### PR DESCRIPTION
**Pull request by isaac73**
# Type of change? 

- [ ] Bug fix
- [x] New feature
- [ ] Enhancement
- [ ] Breaking change (breaks backward compatibility)

# Description, Context, Motivation 
Adding Deployment Details Page
...

**Checklist:**

- [x] All tests passed
- [ ] The code follows the code style and conventions of this project
- [ ] The change requires a change to the documentation
- [ ] I have updated the documentation accordingly

---
This change is [Reviewable](https://reviewable.io/reviews/azure/pcs-remote-monitoring-webui/1112)

### CODE REVIEW

File: src/store/reducers/deploymentsReducer.js [hunk 1]
```diff
@@ -89,24 +121,57 @@ const deleteDeploymentReducer = (state, { payload, fromAction }) => {
...
-    entities: { $set: deployments },
+    entities: { deployments: { $set: deployments } },
     items: { $set: result },
     lastUpdated: { $set: moment() },
     ...setPending(fromAction.type, false)
   });
 };
 
+const updateDeploymentReducer = (state, { payload, fromAction }) => {
+  const deviceStatuses = dot.pick('deviceStatuses', payload);
```
Comment (mechaffin) [inline src/store/reducers/deploymentsReducer.js]: Is dot.pick needed for a single level?

File: src/store/reducers/deploymentsReducer.js [hunk 2]
```diff
@@ -89,24 +121,57 @@ const deleteDeploymentReducer = (state, { payload, fromAction }) => {
...
+  return update(state, {
+    currentDeployment: { $set: payload },
+    deviceStatuses: { $set: deviceStatuses },
+    ...setPending(fromAction.type, false)
+  });
+};
+
+const updateDeployedDevicesReducer = (state, { payload: [modules, devices], fromAction }) => {
+
```
Comment (mechaffin) [inline src/store/reducers/deploymentsReducer.js]: extra line

File: public/locales/en/translations.json [hunk 3]
```diff
@@ -565,13 +565,41 @@
         "priorityPlaceHolder": "Enter priority",
         "namePlaceHolder": "Enter name",
         "targetText": "targeted devices",
-        "infoText": "* Edge Manifest packages that you deploy will be applied to every edge device group in the device group and to any edge group that you add to the device group in the future. The deployment will run continously, verifying that current and future devices are configured correctly (and automatically updating that aren't). Package wont be applied to non-edge devices.",
+        "infoText": "* This deployment runs continuously. Every edge device (and any you add in the future) in the selected device group will receive this package.",
```
[the remark this answers is not in the document]
Comment (isaac73) [reply public/locales/en/translations.json]: No. This is supposed to be red color star. I'm planning to work on styling for messages on the next iteration.

File: public/locales/en/translations.json [hunk 4]
```diff
@@ -565,13 +565,41 @@
...
+      "delete": {
+        "contextMenuName": "Delete",
+        "title": "Delete Deployment?",
+        "delete": "Delete",
+        "cancel": "Cancel",
+        "info": "Deleting {deploymentName} will stop this deployment from being applied to these devices. It may result in a lower priority deployment being applied"
+      }
+    },
+    "details": {
+      "deploymentName": " DEPLOYMENT NAME",
```
[the remark this answers is not in the document]
Comment (isaac73) [reply public/locales/en/translations.json]: done

File: src/components/pages/deployments/deploymentDetails/deploymentDetails.js [hunk 5]
```diff
@@ -0,0 +1,180 @@
+// Copyright (c) Microsoft. All rights reserved.
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/deployments/deploymentDetails/deploymentDetails.js]: We have not done the same for maintenance. On the high level deployments folder contains all the contents that are shown in that tab.

File: src/components/pages/deployments/deploymentDetails/deploymentDetails.js [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+      deltaRowDataMode: true,
+      enableSorting: true,
+      unSortIcon: true,
+      domLayout: 'autoHeight'
+    };
+
+    return (
+      <ComponentArray className="deployments-details-container">
+
+        {isPending && <Indicator key="deployment-details-pending" />}
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/deployments/deploymentDetails/deploymentDetails.js]: removed

File: src/components/pages/deployments/deploymentDetails/deploymentDetails.js [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+      columnDefs: translateColumnDefs(this.props.t, this.columnDefs),
+      sizeColumnsToFit: true,
+      deltaRowDataMode: true,
+      enableSorting: true,
+      unSortIcon: true,
+      domLayout: 'autoHeight'
+    };
+
+    return (
+      <ComponentArray className="deployments-details-container">
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/deployments/deploymentDetails/deploymentDetails.js]: removed

File: src/components/pages/deployments/deploymentDetails/deploymentDetails.js [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+        }
+
+        {
+          !isPending &&
+          <div className="deployment-details-summary-container" key="deployment-details-summary">
+            <div className="deployment-details-summary-labels">
+              {t('deployments.details.deploymentName')}
+            </div>
+            <div className="deployment-name">
+              {currentDeployment.name}
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/deployments/deploymentDetails/deploymentDetails.js]: The glimmer is only used to differentiate an item inside a grid. I don't think it is needed here.

Regarding pageTitle, we have to anyway update the same in maintenance page, we'll update this along with it.

File: src/components/pages/deployments/deploymentDetails/deploymentDetails.js [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+                    {t('deployments.details.priority')}
+                  </div>
+                  <div className="deployment-details-summary-values">
+                    {currentDeployment.priority}
+                  </div>
+                </StatGroup>
+              </StatGroup>
+              <StatGroup className="summary-container-columns">
+                <StatProperty
+                  value={currentDeployment.failedCount}
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/deployments/deploymentDetails/deploymentDetails.js]: done

File: src/components/pages/deployments/deploymentDetails/deploymentDetails.scss [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+      @include rem-fallback(padding, 10px, 24px, 0px, 0px);
+    }
+
+    .deployment-details-summary-labels {
+      @include rem-font-size(12px);
+      @include rem-fallback(padding-bottom, 10px);
+    }
+
+    .stat-failed {
+      @include rem-font-size(16px);
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/deployments/deploymentDetails/deploymentDetails.scss]: fixed

File: src/components/pages/deployments/deploymentDetails/deploymentDetails.scss [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  .deployment-details-devices-affected {
+    @include rem-fallback(border-top, 1px, solid);
+    @include rem-fallback(padding, 35px, 0px, 20px, 0px);
+    @include rem-fallback(margin-top, 50px);
+  }
+
+  @include themify($themes) {
+
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/deployments/deploymentDetails/deploymentDetails.scss]: fixed

File: src/components/pages/deployments/deploymentDetails/deploymentDetails.js [hunk 12]
```diff
@@ -0,0 +1,180 @@
+// Copyright (c) Microsoft. All rights reserved.
+import React, { Component } from 'react';
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/deployments/deploymentDetails/deploymentDetails.js]: done

File: src/components/pages/deployments/deployments.js [hunk 13]
```diff
@@ -55,42 +58,71 @@ export class Deployments extends Component {
...
+                {!!error && <AjaxError t={t} error={error} />}
+                {!error && <DeploymentsGrid {...gridProps} />}
+                {this.state.openFlyoutName === 'newDeployment' && <DeploymentNewContainer t={t} onClose={this.closeFlyout} />}
+              </PageContent>
+            </ComponentArray>
+          } />
+        <Route exact path={'/deployments/:id'}
+          render={(routeProps) =>
+            <ComponentArray>
+              <ContextMenu key="deployment-details-context-menu">
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/deployments/deployments.js]: removed

File: src/components/pages/deployments/deployments.js [hunk 14]
```diff
@@ -55,42 +58,71 @@ export class Deployments extends Component {
...
+                <ContextMenuAlign>
+                  <RefreshBar refresh={fetchDeployments} time={lastUpdated} isPending={isPending} t={t} />
+                  <Protected permission={permissions.createDevices}>
+                    <Btn svg={svgs.trash} onClick={this.openDeleteDeploymentFlyout}>{t('deployments.flyouts.delete.contextMenuName')}</Btn>
+                  </Protected>
+                </ContextMenuAlign>
+              </ContextMenu>
+              <PageContent className="deployments-details-container" key="deployments-page-content">
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/deployments/deployments.js]: removed

File: src/components/pages/maintenance/summary/summary.scss [hunk 15]
```diff
@@ -11,6 +11,10 @@
     @include rem-font-size(20px);
   }
 
+  .summary-stat-container {
+    @include rem-fallback(padding-bottom, 20px);
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/maintenance/summary/summary.scss]: done

File: src/services/models/iotHubManagerModels.js [hunk 16]
```diff
@@ -202,3 +207,13 @@ export const toDeploymentRequestModel = (deploymentModel = {}) => ({
   Priority: deploymentModel.priority,
   Type: deploymentModel.type
 });
+
+export const toEdgeAgentModel = (edgeAgent = {}) => reshape(edgeAgent, {
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/services/models/iotHubManagerModels.js]: fixed

File: src/store/reducers/deploymentsReducer.js (hunk 1 above)
Comment (isaac73) [reply src/store/reducers/deploymentsReducer.js]: It looks better than a ternary operator.

File: src/store/reducers/deploymentsReducer.js (hunk 2 above)
Comment (isaac73) [reply src/store/reducers/deploymentsReducer.js]: fixed

File: src/components/pages/deployments/deploymentDetails/deploymentDetails.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/deployments/deploymentDetails/deploymentDetails.js]: done

File: src/components/app.js [hunk 17]
```diff
@@ -61,7 +61,7 @@ class App extends Component {
       },
       {
         to: '/deployments',
-        exact: true,
+        exact: false,
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/app.js]: done

File: src/components/pages/deployments/deployments.js [hunk 18]
```diff
@@ -55,42 +58,71 @@ export class Deployments extends Component {
...
-        </ContextMenu>
-        <PageContent className="deployments-page-container">
-          <RefreshBar refresh={fetchDeployments} time={lastUpdated} isPending={isPending} t={t} />
-          <PageTitle className="deployments-title" titleValue={t('deployments.title')} />
-          {!!error && <AjaxError t={t} error={error} />}
-          {!error && <DeploymentsGrid {...gridProps} />}
-          {this.state.openFlyoutName === 'newDeployment' && <DeploymentNewContainer t={t} onClose={this.closeFlyout} />}
-        </PageContent>
-      </ComponentArray>
+      <Switch>
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/deployments/deployments.js]: done

File: src/components/pages/deployments/deploymentDetails/deploymentDetails.js [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+              <StatGroup className="summary-container-columns">
+                <StatGroup>
+                  <div className="deployment-details-summary-labels">
+                    {t('deployments.details.deviceGroup')}
+                  </div>
+                  <div className="deployment-details-summary-values">
+                    {currentDeployment.deviceGroupId}
+                  </div>
+                </StatGroup>
+                <StatGroup className="summary-container-second-row">
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/deployments/deploymentDetails/deploymentDetails.js]: I will revisit stats on next iteration

File: src/components/pages/deployments/deploymentDetails/deploymentDetails.js [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+                    size="small" />
+                  <StatProperty
+                    value={pendingCount}
+                    label={t('deployments.details.pending')}
+                    size="small" />
+                </StatGroup>
+              </StatGroup>
+              <StatGroup className="summary-container-columns">
+                <StatGroup>
+                  <div className="deployment-details-summary-labels">
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/deployments/deploymentDetails/deploymentDetails.js]: I will revisit stats on next iteration

File: src/assets/icons/failed.svg [hunk 21]
```diff
@@ -0,0 +1,12 @@
+<!-- Copyright (c) Microsoft. All rights reserved. -->
+
+<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 12.001">
+<g id="Group_5518"  transform="translate(-830.195 -350.613)">
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/assets/icons/failed.svg]: removed

File: src/components/pages/deployments/deploymentDetails/deploymentDetails.js [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+          }
+
+          <h4 className="deployment-details-devices-affected">
+            {t('deployments.details.devicesAffected')}
+          </h4>
+
+          {isDeployedDevicesPending && <Indicator />}
+
+          {
+            deployedDevicesError
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/deployments/deploymentDetails/deploymentDetails.js]: fixed

File: src/components/pages/deployments/deploymentsHome/deployments.scss [hunk 23]
```diff
@@ -4,7 +4,7 @@
 @import 'src/styles/mixins';
 @import 'src/styles/themes';
 
-.deployments-page-container {
+.deployments-page-container, .deployments-details-container {
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/deployments/deploymentsHome/deployments.scss]: fixed

File: src/components/shared/pageStats/statProperty/statProperty.scss [hunk 24]
```diff
@@ -7,7 +7,9 @@
   display: flex;
   flex-flow: row wrap;
   flex-shrink: 0;
-
+  line-height: .85em;
+  align-items: baseline;
+  @include rem-fallback(padding-top, 10px);
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/shared/pageStats/statProperty/statProperty.scss]: done

File: src/store/reducers/deploymentsReducer.js [hunk 25]
```diff
@@ -89,24 +121,57 @@ const deleteDeploymentReducer = (state, { payload, fromAction }) => {
...
+const updateDeployedDevicesReducer = (state, { payload: [modules, devices], fromAction }) => {
+  const normalizedDevices = normalize(devices, deployedDevicesListSchema).entities.deployedDevices;
+  const normalizedModules = normalize(modules, deployedDevicesListSchema).entities.deployedDevices;
+  const deployedDevices = Object.keys(normalizedDevices).reduce((acc, deviceId) => ({
+    ...acc,
+    [deviceId]: {
+      ...(acc[deviceId] || {}),
+      firmware: normalizedDevices[deviceId].firmware
+    }
+  }), normalizedModules
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/store/reducers/deploymentsReducer.js]: I did my best

File: src/store/reducers/deploymentsReducer.js [hunk 26]
```diff
@@ -89,24 +121,57 @@ const deleteDeploymentReducer = (state, { payload, fromAction }) => {
...
+    currentDeploymentLastUpdated: { $set: moment() },
+    ...setPending(fromAction.type, false)
+  });
+};
+
+const updateDeployedDevicesReducer = (state, { payload: [modules, devices], fromAction }) => {
+  const normalizedDevices = normalize(devices, deployedDevicesListSchema).entities.deployedDevices;
+  const normalizedModules = normalize(modules, deployedDevicesListSchema).entities.deployedDevices;
+  const deployedDevices = Object.keys(normalizedDevices)
+    .reduce(
```
Comment (spryor) [inline src/store/reducers/deploymentsReducer.js]: This is more what I had in mind. A clear distinction between when the function opens, closes and there the parameters are.
```javascript
.reduce(
  (acc, deviceId) => ({
    ...acc,
    [deviceId]: { 
      ...(acc[deviceId] || {}), 
      firmware: normalizedDevices[deviceId].firmware 
    }
  }),
  normalizedModules
);
```

File: src/components/pages/deployments/deployments.test.js [hunk 27]
```diff
@@ -4,7 +4,7 @@ import React from 'react';
 import { shallow } from 'enzyme';
 import 'polyfills';
 
-import { Deployments } from './deployments';
+import { Deployments } from './deploymentsHome/deployments';
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/deployments/deployments.test.js]: added

File: src/components/pages/deployments/deploymentsHome/deployments.js [hunk 28]
```diff
@@ -29,7 +29,8 @@ export class Deployments extends Component {
     super(props);
     this.state = {
       ...closedFlyoutState,
-      contextBtns: null
+      contextBtns: null,
+      softSelectedDeployment: undefined
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/deployments/deploymentsHome/deployments.js]: removed

File: src/services/iotHubManagerService.js [hunk 29]
```diff
@@ -74,6 +75,24 @@ export class IoTHubManagerService {
...
   }
 
+  /** Returns deployment */
+  static getDeployment(id) {
+    return HttpClient.get(`${ENDPOINT}deployments/${id}?includeDeviceStatus=true`)
+      .map(toDeploymentModel);
+  }
+
+  /** Queries EdgeAgent */
+  static queryEdgeAgents(query) {
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/services/iotHubManagerService.js]: updated the name

**Pull request merged**